loopd/swapclient_server.go

```diff
@@ -992,9 +1003,73 @@ func validateLoopOutRequest(chainParams *chaincfg.Params, confTarget int32,
 	}
 
 	// Check that the label is valid.
-	if err := labels.Validate(label); err != nil {
+	if err := labels.Validate(req.Label); err != nil {
+		return 0, err
+	}
+
+	channels, err := lnd.ListChannels(ctx)
+	if err != nil {
 		return 0, err
 	}
 
-	return validateConfTarget(confTarget, loop.DefaultSweepConfTarget)
+	unlimitedChannels := len(req.OutgoingChanSet) == 0
+	outgoingChanSetMap := make(map[uint64]bool)
+	for _, c := range req.OutgoingChanSet {
+		outgoingChanSetMap[c] = true
+	}
+
+	var activeChannelSet []lndclient.ChannelInfo
+	for _, c := range channels {
+		// Don't bother looking at inactive channels.
+		if !c.Active {
+			continue
+		}
+
+		// If no outgoing channel set was specified then all active
+		// channels are considered. However, if a channel set was
+		// specified then only the specified channels are considered.
+		if unlimitedChannels || outgoingChanSetMap[c.ChannelID] {
+			activeChannelSet = append(activeChannelSet, c)
+		}
+	}
+
+	// Determine if the loop out request is theoretically possible given
+	// the amount requested, the maximum possible routing fees,
+	// the available channel set and the fact that equal splitting is
+	// used for MPP.
+	requiredBalance := btcutil.Amount(req.Amt + req.MaxSwapRoutingFee)
+	if !isRoutable(activeChannelSet, requiredBalance, maxParts) {
+		return 0, fmt.Errorf("%w: Requested swap amount of %d "+
+			"sats along with the maximum routing fee of %d sats "+
+			"is more than what can be routed given current state "+
+			"of the channel set", errBalanceTooLow, req.Amt,
+			req.MaxSwapRoutingFee)
+	}
+
+	return validateConfTarget(
+		req.SweepConfTarget, loop.DefaultSweepConfTarget,
+	)
+}
+
+// isRoutable checks if a given amount can be looped out given a specific
+// channel set and a maximum number of MPP parts that can be used.
+func isRoutable(channels []lndclient.ChannelInfo, amt btcutil.Amount,
+	maxParts uint32) bool {
+
+	var parts uint32
+	for parts = 1; parts <= maxParts; parts++ {
+		minHTLC := amt.MulF64(1 / float64(parts))
+
+		var totalRoutableParts uint32
+		for _, c := range channels {
+			routableParts := (c.LocalBalance / minHTLC).ToUnit(btcutil.AmountSatoshi)
```

Comment:
Thanks so much for taking the time to create the example! Very very helpful. Yeah I didn't quite grasp the splitting in half logic before but I understand now 👍  
(The only thing is, I think line 38 should be `split /= 2` right? Otherwise an infinite loop can occur)

Reply:
Yes indeed, that was a mistake, I didn't test it well enough :)